paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala

```diff
@@ -0,0 +1,199 @@
+              }
+              write.prepareCommit().asScala.map(serializer.serialize)
+            } finally {
+              write.close()
+            }
+        }
+        .collect()
+        .map(deserializeCommitMessage(serializer, _))
+
+    try {
+      val tableCommit = writeBuilder.newCommit().asInstanceOf[InnerTableCommit]
+      tableCommit.commit(committables.toList.asJava)
+    } catch {
+      case e: Throwable => throw new RuntimeException(e);
+    }
+
+    Seq.empty
+  }
+
+  override def withNewChildrenInternal(newChildren: IndexedSeq[LogicalPlan]): LogicalPlan =
+    this.asInstanceOf[WriteIntoPaimonTable]
+}
+
+object WriteIntoPaimonTable {
+
+  sealed trait BucketProcessor {
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row]
+  }
+
+  case class CommonBucketProcessor(
+      writeBuilder: BatchWriteBuilder,
+      bucketColIndex: Int,
+      toRow: ExpressionEncoder.Serializer[Row],
+      fromRow: ExpressionEncoder.Deserializer[Row])
+    extends BucketProcessor {
+
+    private val rowType = writeBuilder.rowType
+
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row] = {
+      val batchTableWrite = writeBuilder.newWrite()
+      new Iterator[Row] {
+        override def hasNext: Boolean = rowIterator.hasNext
+
+        override def next(): Row = {
+          val row = rowIterator.next
+          val sparkInternalRow = toRow(row)
+          sparkInternalRow.setInt(
+            bucketColIndex,
+            batchTableWrite.getBucket(new SparkRow(rowType, row)))
+          fromRow(sparkInternalRow)
+        }
+      }
+    }
+  }
+
+  case class DynamicBucketProcessor(
+      fileStoreTable: FileStoreTable,
+      rowType: RowType,
+      bucketColIndex: Int,
+      toRow: ExpressionEncoder.Serializer[Row],
+      fromRow: ExpressionEncoder.Deserializer[Row]
+  ) extends BucketProcessor {
+
+    private val targetBucketRowNumber = fileStoreTable.coreOptions.dynamicBucketTargetRowNum
+
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row] = {
+      val indexFileHandler = fileStoreTable.store.newIndexFileHandler
+      val partitionIndex = mutable.Map.empty[BinaryRow, PartitionIndex]
+      val rowPartitionKeyExtractor = new RowPartitionKeyExtractor(fileStoreTable.schema)
+      new Iterator[Row]() {
+        override def hasNext: Boolean = rowIterator.hasNext
+
+        override def next(): Row = {
+          val row = rowIterator.next
+          val sparkRow = new SparkRow(rowType, row)
+          val hash = rowPartitionKeyExtractor.trimmedPrimaryKey(sparkRow).hashCode
+          val partition = rowPartitionKeyExtractor.partition(sparkRow)
+          val index = partitionIndex.getOrElseUpdate(
+            partition,
+            PartitionIndex.loadIndex(
+              indexFileHandler,
+              partition,
+              targetBucketRowNumber,
+              (_) => true))
```

Comment:
Here just return true, maybe is not enough.
For example, if this is an empty table, then all tasks will assign the new data to bucket-0.
In fact, the amount of data inside bucket-0 may be close to targetBucketRowNumber * taskNumber.